src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+    fn create(log_segment_path: &Path, trace_path: &Path) -> Result<Self, anyhow::Error> {
+        let messages = read_segment(log_segment_path)?;
+        Batch::create_from_messages(messages, trace_path, None)
+    }
+
+    /// Read in and consolidate a list of messages, and write them to a new batch file
+    /// in `trace_path`.
+    ///
+    /// Will also compact updates up to `compaction_frontier` if provided..
+    fn create_from_messages(
+        messages: Vec<Message>,
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut upper: Option<Timestamp> = None;
+        let mut lower: Option<Timestamp> = None;
+        let mut time_data = BTreeMap::new();
+        for message in messages.iter() {
+            match message {
+                Message::Progress(time) => match (lower, upper) {
+                    (None, None) => {
+                        lower = Some(*time);
+                    }
+                    (Some(l), None) => {
+                        assert!(*time >= l);
+                        upper = Some(*time);
+                    }
+                    (Some(_), Some(u)) => {
+                        assert!(*time >= u);
+                        upper = Some(*time);
+                    }
+                    (None, Some(_)) => unreachable!(),
+                },
+                Message::Data(Update {
+                    row,
+                    timestamp,
+                    diff,
+                }) => {
+                    let time = if let Some(frontier) = compaction_frontier {
+                        std::cmp::min(frontier, *timestamp)
+                    } else {
+                        *timestamp
+                    };
+                    let entry = time_data.entry((time, row)).or_insert(0);
+                    *entry += diff;
+                }
+            }
+        }
+
+        // Now let's prepare the output
+        let mut buf = Vec::new();
+        assert!(lower.is_some());
+        assert!(upper.is_some());
+
+        let lower = lower.unwrap();
+        let upper = upper.unwrap();
+
+        // Frame each batch with its lower and upper bound timestamp.
+        // TODO: match the behavior of CDCv2 updates with a count of messages
+        // at each timestamp.
+        encode_progress(lower, &mut buf)?;
+        for ((timestamp, row), diff) in time_data.into_iter() {
+            if diff == 0 {
+                continue;
+            }
+
+            encode_update(row, timestamp, diff, &mut buf)?;
+        }
+
+        encode_progress(upper, &mut buf)?;
+
+        let batch_name = format!("batch-{}-{}", lower, upper);
+        let batch_path = trace_path.join(&batch_name);
+        let batch_tmp_path = trace_path.join(format!("{}-tmp", batch_name));
+        // Write the file first suffixed with "-tmp" and then rename to guard against
+        // partial writes.
+        fs::write(&batch_tmp_path, buf)
```

Comment:
`fs::write` will not call `sync_all()`, I think we need to do it manually

Reply:
oof thank you for the great catch.